akka-remote/src/main/scala/akka/remote/artery/SendQueue.scala

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.artery
+
+import java.util.Queue
+import akka.stream.stage.GraphStage
+import akka.stream.stage.OutHandler
+import akka.stream.Attributes
+import akka.stream.Outlet
+import akka.stream.SourceShape
+import akka.stream.stage.GraphStageLogic
+import org.agrona.concurrent.ManyToOneConcurrentArrayQueue
+import akka.stream.stage.GraphStageWithMaterializedValue
+import org.agrona.concurrent.ManyToOneConcurrentLinkedQueueTail
+import org.agrona.concurrent.ManyToOneConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicInteger
+import scala.annotation.tailrec
+
+/**
+ * INTERNAL API
+ */
+private[remote] object SendQueue {
+  trait QueueValue[T] {
+    def inject(queue: Queue[T]): Unit
+    def offer(message: T): Boolean
+  }
+
+  private trait WakeupSignal {
+    def wakeup(): Unit
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] final class SendQueue[T] extends GraphStageWithMaterializedValue[SourceShape[T], SendQueue.QueueValue[T]] {
+  import SendQueue._
+
+  val out: Outlet[T] = Outlet("SendQueue.out")
+  override val shape: SourceShape[T] = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, QueueValue[T]) = {
+    @volatile var needWakeup = false
+    @volatile var queue: Queue[T] = null
+    // TODO perhaps try similar with ManyToOneConcurrentLinkedQueue or AbstractNodeQueue
+
+    val logic = new GraphStageLogic(shape) with OutHandler with WakeupSignal {
+      private val wakeupCallback = getAsyncCallback[Unit] { _ ⇒
+        if (isAvailable(out))
+          tryPush()
+      }
+
+      override def preStart(): Unit = {
+        needWakeup = true
+      }
+
+      override def onPull(): Unit = {
+        if (queue ne null)
+          tryPush()
+      }
+
+      @tailrec private def tryPush(firstAttempt: Boolean = true): Unit = {
+        queue.poll() match {
+          case null ⇒
+            needWakeup = true
+            // additional poll() to grab any elements that might missed the needWakeup
+            // and have been enqueued just after it
+            if (firstAttempt)
+              tryPush(firstAttempt = false)
```

Comment:
performance is still not great though, might be that we are thrashing the cache. Might be that the queue is mostly empty and hence much contention between the reader and writer?